### PR TITLE
Enable deleting all hosts using cautionary args

### PR DESCRIPTION
**Pull request by thearifismail**
# Overview

This PR is being created to address [ESSNTL-1967](https://issues.redhat.com/browse/ESSNTL-1967).
(A description of your PR's changes, along with why/context to the PR, goes here.)

## PR Checklist

- [x] Keep PR title short, ideally under 72 characters
- [x] Descriptive comments provided in complex code blocks
- [x] Tests: validate optimal/expected output
- [x] Tests: validate exceptions and failure scenarios
- [ ] Tests: edge cases
- [ ] Recovers or fails gracefully during potential resource outages (e.g. DB, Kafka)
- [ ] Documentation, if this PR changes the way other services interact with host inventory
- [ ] Links to related PRs

## Secure Coding Practices Documentation Reference

You can find documentation on this checklist [here](https://github.com/RedHatInsights/secure-coding-checklist).

## Secure Coding Checklist

- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

**Comment by ryandillinfelton** [comment]
This seems quite safe, but I can't help but feel it may be more safe to make a separate delete_all endpoint.
Maybe
`/api/delete/all` 
or
`/api/delete-all/`
for even more separation. I would still support a confirmation parameter.

One reason is, with how destructive the action is, I think the more idiot-proof we can make it the better. Having it be an entirely separate endpoint exclusively for deleting everything makes me feel safer. No need to ever touch it at all if I don't want all my hosts gone.

Another is that syntactically I think it's a bit weird. The query parameter isn't modifying the action, it's asking for a different (admittedly similar) action. As a result the possible query parameters interact is odd ways. For example:
`/api/delete?providerId={some_id}&delete_all=true&confirm=true`
is valid, but the filter is irrelevant.

I also worry someone might think they need to add `delete_all` to have all the hosts selected by their query deleted, or that they must always attach `confirm` to their query as if it's a protection for any delete, not just `delete_all` which would reduce it's effectiveness. I think the latter point could be avoided by renaming it to `confirm_delete_all` to be more explicit about what it's confirming.

My personal preference is for `api/delete-all?confirm_delete_all=true` as it seems the safest, and avoids syntactic weirdness since the only possible query parameter would be the confirmation. Then the `/delete` endpoint would be reserved for deleting hosts through filters, while `delete/all` or `/delete-all` would be reserved for clearing all the hosts from an account. `confirm` would also be acceptable in this case, but I'm all for making this as cumbersome and explicit as possible a la typing the whole repo name when deleting a repo on github.

**Comment by kruai** [comment]
> This seems quite safe, but I can't help but feel it may be more safe to make a separate delete_all endpoint. Maybe `/api/delete/all` or `/api/delete-all/` for even more separation. I would still support a confirmation parameter.
> 
> One reason is, with how destructive the action is, I think the more idiot-proof we can make it the better. Having it be an entirely separate endpoint exclusively for deleting everything makes me feel safer. No need to ever touch it at all if I don't want all my hosts gone.
> 
> Another is that syntactically I think it's a bit weird. The query parameter isn't modifying the action, it's asking for a different (admittedly similar) action. As a result the possible query parameters interact is odd ways. For example: `/api/delete?providerId={some_id}&delete_all=true&confirm=true` is valid, but the filter is irrelevant.
> 
> I also worry someone might think they need to add `delete_all` to have all the hosts selected by their query deleted, or that they must always attach `confirm` to their query as if it's a protection for any delete, not just `delete_all` which would reduce it's effectiveness. I think the latter point could be avoided by renaming it to `confirm_delete_all` to be more explicit about what it's confirming.
> 
> My personal preference is for `api/delete-all?confirm_delete_all=true` as it seems the safest, and avoids syntactic weirdness since the only possible query parameter would be the confirmation. Then the `/delete` endpoint would be reserved for deleting hosts through filters, while `delete/all` or `/delete-all` would be reserved for clearing all the hosts from an account. `confirm` would also be acceptable in this case, but I'm all for making this as cumbersome and explicit as possible a la typing the whole repo name when deleting a repo on github.

These are valid points. I'd prefer not to add yet another DELETE `/hosts` endpoint, but I do completely understand your concern... Here are my thoughts on these approaches, for what they're worth.

If we stick with having one DELETE endpoint:
- We could raise an exception if both a filter and `delete_all` are provided
- In addition to the requirements we've already discussed, I think having very clear descriptions for these params in our API spec should be sufficient idiot-proofing for developers poking around on the API. The average user will be deleting from the UI, which will not be able to make this mistake.

If we make a separate DELETE endpoint:
- Since the current endpoint is DELETE `/hosts/{host_id_list}`, I'd prefer the new endpoint to at least follow our established pattern and be something like DELETE `/hosts/all?confirm_delete_all=true`

**Comment by thearifismail** [comment]
Though, I was in favor of creating a new endpoint for deleting all hosts, @kruai argument is persuasive.  It is simpler than adding a new endpoint.  I am leaning toward raising an exception when `delete_all` is used with an other filter.  The second point in Asa's response to Ryan is already in place

**Review by kruai — CHANGES_REQUESTED** [review]

File: api/host.py [hunk 1]
```diff
@@ -132,6 +133,41 @@ def get_host_list(
     return flask_json_response(json_data)
 
 
+def _validate_filter(**kwargs):
```
[the remark this answers is not in the document]
Comment (thearifismail) [reply api/host.py:136]: renamed the function to `_delete_all_hosts_filter()`

File: api/host.py [hunk 2]
```diff
@@ -146,30 +182,46 @@ def delete_host_list(
...
+        hostname_or_id=hostname_or_id,
+        insights_id=insights_id,
+        provider_id=provider_id,
+        provider_type=provider_type,
+        registered_with=registered_with,
+        tags=tags,
+        filter=filter,
+        delete_all=delete_all,
+        confirm=confirm,
+    ):
```
[the remark this answers is not in the document]
Comment (thearifismail) [reply api/host.py]: Using `args = locals()`, causes test failures, which I think puts `staleness` into action.  Using locals() should wait until after staleness defaults have been removed.

File: swagger/api.spec.yaml [hunk 3]
```diff
@@ -663,6 +665,20 @@ components:
...
+      schema:
+        type: boolean
+      description: The option to delete all hosts in an account.
+      required: false
+    confirm:
+      in: query
+      name: confirm
+      schema:
+        type: boolean
+      description: The Confirm option provided for additional caution
```
[the remark this answers is not in the document]
Comment (thearifismail) [reply swagger/api.spec.yaml]: Updated the description to this suggestion.

File: tests/test_api_hosts_delete.py [hunk 4]
```diff
@@ -143,6 +143,69 @@ def test_delete_hosts_using_filter(
...
+):
+    created_hosts = db_create_multiple_hosts(how_many=len(XJOIN_HOSTS_RESPONSE_FOR_FILTERING["hosts"]["data"]))
+    host_ids = [str(host.id) for host in created_hosts]
+
+    # set the new host ids in the xjoin search reference.
+    resp = deepcopy(XJOIN_HOSTS_RESPONSE_FOR_FILTERING)
+    ind = 0
+    for id in host_ids:
+        resp["hosts"]["data"][ind]["id"] = id
+        ind += 1
```
[the remark this answers is not in the document]
Comment (thearifismail) [reply tests/test_api_hosts_delete.py]: updated

File: tests/test_api_hosts_delete.py [hunk 5]
```diff
@@ -143,6 +143,69 @@ def test_delete_hosts_using_filter(
...
+
+    # Make the new hosts available in xjoin-search to make them available
+    # for querying for deletion using filters
+    patch_xjoin_post(response, status=200)
+
+    # delete all hosts using incomplete filter
+    response_status, response_data = api_delete_filtered_hosts({"confirm": True})
+
+    assert_response_status(response_status, expected_status=400)
+    assert "Deleting every host requires delete_all=true&confirm=true" in response_data["detail"]
```
[the remark this answers is not in the document]
Comment (thearifismail) [reply tests/test_api_hosts_delete.py]: parametrize the filter to use confirm and delete_all

File: tests/test_xjoin.py [hunk 6]
```diff
@@ -1739,6 +1739,14 @@ def test_xjoin_search_query_using_hostfilter_provider(
     )
 
 
+def test_xjoin_search_query_to_delete_all_hosts(mocker, graphql_query_empty_response, api_delete_filtered_hosts):
+    query_params = {"delete_all": True, "confirm": True}
+    api_delete_filtered_hosts(query_params)
+
+    # using mocker.ANY because no filter is used to fetch all hosts.
+    graphql_query_empty_response.assert_called_once_with(HOST_IDS_QUERY, {"filter": (mocker.ANY)}, mocker.ANY)
```
[the remark this answers is not in the document]
Comment (thearifismail) [reply tests/test_xjoin.py:1747]: Thanks to the default values for staleness `mocker.ANY` is needed.  Will fix updated when `args = locals()` will be added.

**Comment by thearifismail** [comment]
This PR work is covered by https://github.com/RedHatInsights/insights-host-inventory/pull/1082

**Pull request closed without merging**